### PR TITLE
Reverse dependency between platform and main

### PR DESCRIPTION
**Pull request by alevy**
~~NOTE: Comment at will, but don't merge yet, something seems to be broken with running multiple apps. Actually there was no bug. The sensors app just required more than was provided to support both the sam4l and nrf51822. So now, waiting on #89 to shift the burden of allocating app memory to the platform (which we want anyway) to circumvent this problem.~~

_Ready to merge now. Loading processes in the NRF51822 doesn't seem to work, but it doesn't work on master either, so... Let's fix that in a separate issue._

In the new world, the `platform` crate depends is the main crate and it
depends on the `main` crate (which includes the scheduler). The platform
defines a reset handler for use by the chip's vector table. The
reset_handler calls the kernel `main` function (i.e. the scheduler)
when it's done initializing the platform.

This required some other cleanup that we wanted anyway, e.g., moving
systick into the architecture crate instead of a platform one, where it
belongs.

Processes are now loaded in platform-specific functions in the platform crate. This is useful for defining different sized memories and number of processes (/cc @lizardo) as well as being able to locate app code on flash in platform specific ways.

### CODE REVIEW

File: src/arch/cortex-m4/mpu.rs [hunk 1]
```diff
@@ -77,25 +78,16 @@ impl MPU {
...
-    /// Sets the base address, size and access attributes of the given MPU
-    /// region number.
-    ///
-    /// `region_num`: an MPU region number 0-7
-    /// `start_addr`: the region base address. Lower bits will be masked
-    ///               according to the region size. 
-    /// `len`       : region size as a function 2^(len + 1)
-    /// `execute`   : whether to enable code execution from this region
-    /// `ap`        : access permissions as defined in Table 4.47 of the user
-    ///               guide.
```
Comment (yuriks) [inline src/arch/cortex-m4/mpu.rs]: What's up with removing the docs?

File: src/platform/storm/Makefile.mk [hunk 2]
```diff
@@ -7,11 +7,30 @@ SDB_VERSION=$(shell git show-ref -s HEAD)
 SDB_NAME=storm.rs
 SDB_DESCRIPTION="An OS for the storm"
 
-$(BUILD_PLATFORM_DIR)/libplatform.rlib: $(call rwildcard,$(SRC_DIR)platform/storm,*.rs) $(BUILD_PLATFORM_DIR)/libcore.rlib $(BUILD_PLATFORM_DIR)/libhil.rlib $(BUILD_PLATFORM_DIR)/libsam4l.rlib $(BUILD_PLATFORM_DIR)/libdrivers.rlib | $(BUILD_PLATFORM_DIR)
+PLATFORM_DEPS=$(BUILD_PLATFORM_DIR)/libcore.rlib $(BUILD_PLATFORM_DIR)/libsupport.rlib
```
Comment (yuriks) [inline src/platform/storm/Makefile.mk]: Use spaces around the operator, possibly use `:=` insead.

File: src/platform/storm/main.rs [hunk 3]
```diff
@@ -32,8 +31,7 @@ pub mod systick;
 static mut spi_read_buf:  [u8; 64] = [0; 64];
 static mut spi_write_buf: [u8; 64] = [0; 64];
 
-pub struct Platform {
-    chip: sam4l::chip::Sam4l,
+pub struct Firestorm {
```
Comment (yuriks) [inline src/platform/storm/main.rs]: Does this still need to be pub now that it's abstracted by the trait?

**Comment by KBaichoo** [comment]
Talked with @alevy a bit about this and thought this might be worth noting.

For the future we might want applications that aren't processes to be supported (i.e. they're on the board, but currently not processes as they're not running). This would allow us to bootup an application as a process later on. 

We could use some memory from the static memory allocated for processes 'MEMORIES' to hold some meta information. We could stick the process meta info at the beginning / end of processes's memory and using subregions to hide the meta info from the process.

**Comment by alevy** [comment]
To expand on @KBaichoo's comment:

It's shouldn't be a problem to have a fixed number of process slots and replace running processes with other ones (there are of course interesting state-saving questions there).

However, one of Kevin's points was we may want to support a dynamic number of processes (e.g., maybe some of them require less RAM than others). In this case, we need to store the running process list in a dynamically sized data structure, as well as have a dynamic number of them. The challenge there is that we don't have a heap.

Naively reusing the process memory block (i.e. `MEMORIES`) as a heap could be problematic since the MPU requires memory regions to be align in particular ways, which would be complicated by fragmentation.

However, since we'd always (probably?) have one-to-one process meta-data and running processes, we can just always allocate the process meta-data at the beginning/end of process memory, and hide it from the process using an MPU subregion.

**Comment by yuriks** [comment]
The classic embedded solution for these "problems" is to simply statically allocate a reasonable number of slots that are enough to fit the number of processes required for our application. Any reason why this is harder/undesirable here?

**Comment by phil-levis** [comment]
Undesirable in that it is fundamentally hacky, in that it either wastes memory or fails dynamically. E.g., it could be you have enough RAM and code space but have run out of slots.

So from an engineering standpoint, a totally reasonable thing to do until we have a better answer. But from a more research perspective of "let's see how far we can push a system based on these principles" it would be good to have a better solution.

**Comment by alevy** [comment]
@yuriks another take on @phil-levis's answer is that in the standard embedded systems world you know the application(s) ahead of time. But that's not _necessarily_ the case with Tock's target platforms -- we can imagine scenarios in which the workload might change from a single large application as planned by the original platform manufacturer to many small ones installed by the end-user.

For example, the Pebble watch supports exactly two applications at a time -- one "background" app and one app the user is interacting with. That seems like a reasonable choice given the constraints of the hardware, but is pretty limiting.

So, in many cases, in fact, it probably makes sense to just allocate a fixed number as we do now. But in some cases, it would be better to avoid that. If the OS can support both (and the platform gets to decide) even better. If this can be done without a performance penalty (e.g. because the processes data structure implements a trait that just goes away at compile time) that's the best!

**Pull request merged**